design/oauth2-device-authorization-grant.md

```diff
@@ -0,0 +1,144 @@
+* **JIRA**: [KEYCLOAK-7675](https://issues.jboss.org/browse/KEYCLOAK-7675)
+
+
+## Motivation
+
+[The OAuth 2.0 Device Authorization Grant](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15) is designed for internet-connected devices that have limited input capabilities or lack a suitable browser. The spec is still draft, but it has already been implemented by many major IdPs. By supporting this spec, we will be able to use Keycloak in more fields.
+
+## Endpoint
+
+### Device Authorization Endpoint
+
+In the spec, Device Authorization Endpoint is introduced for authorization request from the device which is an OAuth client. Keycloak has common AuthorizationEndpointBase class which provides common logic like checking realm, SSL and so on. We implement this new endpoint using the common base class. The URL is defined as `/realms/{realmName}/openid-connect/device/auth` in current implementation proposal.
+
+### Verification Endpoint
+
+Keycloak must provide a new endpoint for the verification process of the user code which is returned by device authorization request. Also, it's good UX that providing a shorter verification URI because the end-user need to enter the URI manually into their browser if the device has a limited display. So, it's defined as `/realms/{realmName}/device` in current implementation proposal.
+
+This endpoint provides user code verification and user authentication, which we will discuss in User Interaction section.
+
+### Token Endpoint
+
+Token request for the spec is represented with new grant type of `urn:ietf:params:oauth:grant-type:device_code`. We add the process of this new grant type into an existing TokenEndpoint class for OAuth 2.0/OIDC.
+
+
+## User Interaction when verifying a user code
+
+The spec says that the details of the user interaction when processing the verification are up to the authorization server in [3.3. User Interaction](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.3). There are two methods for the process which are imeplemented by major IdPs. *Login First* or *User Code Verification First*.
+
+### Login First
+
+In this case, the end-user need to login first before entering a user code in the user interaction. It means the authorization server must provide a shared authentication flow because the server can't detect the client ID yet which is bind to the user code. However, the risk getting user code brute force attack is low as attackers must log in first.
+
+### User Code Verification First
+
+In this case, the end-user need to enter a user code first in the user interaction. It means the authorization server can provide different authentication flow per client because the server can detect the client ID from the user code. However, the risk getting user code brute force attack is higher than Login First as attackers can try to enter user codes freely.
+
+### Which user interaction does keycloak should support?
```

Comment:
I agree that user code first is what we should consider. However, if the user is using the `verification_uri_complete` we can prompt for credentials beforehand given that the `user_code` is already available from the URL. 

Are you considering Non-textual Verification URI Optimization (`verification_uri_complete`) in this proposal?

Reply:
Regarding rate limiting, I guess we can reuse some stuff that we use in Security Defenses/Brute Force ?

Reply:
verification_uri_complete should be implemented as it allows displaying a QR code with the link and not require users to enter user code manually. It should show a first page asking the user to consent to linking the device prior to continuing (this page should be displayed first in place of the page asking for the user code with verification_uri)

Reply:
Thank you for feedback. I'll add an explanation of authentication to my proposal. In my proposal, the regular browser flow is used for the user authentication after user code verification.

> verification_uri_complete 

In [3.3.1](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.3.1), the spec says `The server SHOULD display the "user_code" to the user and ask them to verify that it matches the "user_code" being displayed on the device, to confirm they are authorizing the correct device.` So I agree that showing a page first in place of the page asking for the user code with verification_uri.

> rate limiting / brute force protection

Unlike the existing Brute Force protection, the user ID is unknown at that time. 
How about using the value of Cookie `AUTH_SESSION_ID` as the key for handling rate limiting / brute force protection? Or should the source IP be the key?

Reply:
I'd say anyone trying to brute force something would clear cookies after each attempt. At least that's what I'd do ;)

Source IP is also a bit questionable as most likely a brute force attack would require significant attempts here, meaning a bot net would be used.

To be honest I'm not sure how best to proceed with this right now, but I don't think the brute force protection we have today can be directly used here and we'd need something different (and smarter?).

Reply:
How about implementing rate limiting user code attempts, which is described in the spec [5.1. User Code Brute Forcing](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-5.1)? I think we can achieve rate limiting with the following logic.

* When attacker enters an invalid user code, an error event is recorded into `EVENT_ENTITY` table.
* If the code is valid, keycloak check whether the attempts exceeds the threshold number of times from the code issued time referring `EVENT_ENTITY` table.
* If the attempt exceeds the limit, reject the verification process of user code.

Reply:
Maybe we should, just in the same way that we control the `interval` which a client can pool the AS for user approval, control the interval that the user can submit the `user_code`. In combination with the code lifetime, this should make life really hard for an attacker.

Beyond that, admins should take into account any network tool to mitigate suspicious behavior.

Reply:
I wonder if we could use the same logic as we do for users. If a client tries an invalid user code we log an attempt against the client. Then we make it configurable in the same way as brute force protection with regards to N attempts, a timeout to purge attempts, disabling the client like we disable users, etc..

It could open up for a DoS attack though as I assume with device grant we are talking about public clients, so would be pretty simple for someone to block the client.

Reply:
> Maybe we should, just in the same way that we control the interval which a client can pool the AS for user approval, control the interval that the user can submit the user_code.

I do not think it can be done the same way. The reason is that the client ID cannot be identified when an invalid UserCode is sent. So [the way I wrote above](https://github.com/keycloak/keycloak-community/pull/6#discussion_r286799277) may be a little tricky. It judges that a DoS attack is hit the valid code if the number of failed attempts exceeds the threshold while the UserCode is issued and verified by transmission from the user.

This method is open to DoS attacks, as @stianst says. A malicious user can easily interfere with UserCode verification of this Device flow.
However, normal logins are open to DoS attacks too, so DoS attackers can intentionally lock the account if they know active username (or email), and we have the same problem. To be honest, I think this kind of DoS attack should be guarded using such as WAF(Web Application Firewall) in front of keycloak instead of keycloak level. And if we use WAF, it should generally have Rate Limiting. In that case, I think that implementing the BruteForce protection of Device flow with keycloak is not effective. What do you think?

Reply:
Yeah, you are right. There is no silver bullet for invalid user code. After talking with @stianst , we are not willing to provide any rate limiting but rely on whatever you are using to protect your app such as a WAF or even an API Manager.